srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1246 @@
+//! - `unreserve`: Move up to an amount from reserved balance to free balance. This function cannot fail.
+//! - `slash`: Deduct up to an amount from the combined balance of `who`, preferring to deduct from the
+//!	free balance. This function cannot fail.
+//! - `reward`: Add up to an amount to the free balance of an account.
+//! - `slash_reserved`: Deduct up to an amount from reserved balance of an account. This function cannot fail.
+//! - `repatriate_reserved`: Move up to an amount from reserved balance of an account to free balance of another
+//! account.
+//! - `check_permission`: Check permission to perform burn, mint or update.
+//! - `ensure_can_withdraw`: Check if the account is able to make a withdrawal of the given amount
+//!	for the given reason.
+//!
+//! ### Usage
+//!
+//! The following examples show how to use the Generic Asset module in your custom module.
+//!
+//! ### Examples from the PRML
+//!
+//! The Fees module uses the `Currency` trait to handle fee charge/refund, and its types inherit from `Currency`:
+//!
+//! ```
+//! use support::{
+//! 	traits::{Currency, ExistenceRequirement, WithdrawReason},
+//! 	dispatch::Result,
+//! };
+//! # pub trait Trait: system::Trait {
+//! # 	type Currency: Currency<Self::AccountId>;
+//! # }
+//! type AssetOf<T> = <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+//!
+//! fn charge_fee<T: Trait>(transactor: &T::AccountId, amount: AssetOf<T>) -> Result {
+//! 	// ...
+//! 	T::Currency::withdraw(
+//!			transactor,
+//!			amount,
+//!			WithdrawReason::TransactionPayment,
+//!			ExistenceRequirement::KeepAlive,
+//!		)?;
+//! 	// ...
+//! 	Ok(())
+//! }
+//!
+//! fn refund_fee<T: Trait>(transactor: &T::AccountId, amount: AssetOf<T>) -> Result {
+//! 	// ...
+//! 	T::Currency::deposit_into_existing(transactor, amount)?;
+//! 	// ...
+//! 	Ok(())
+//! }
+//!
+//! # fn main() {}
+//! ```
+//!
+//! ## Genesis config
+//!
+//! The Generic Asset module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use parity_codec::{Decode, Encode, HasCompact};
+
+use primitives::traits::{
+	CheckedAdd, CheckedSub, MaybeSerializeDebug, Member, One, Saturating, SimpleArithmetic, Zero,
+};
+
+use rstd::prelude::*;
+use rstd::{cmp, result};
+use support::dispatch::Result;
+use support::{
+	decl_event, decl_module, decl_storage, ensure,
+	traits::{
+		Currency, ExistenceRequirement, Imbalance, LockIdentifier, LockableCurrency, ReservableCurrency,
+		SignedImbalance, UpdateBalanceOutcome, WithdrawReason, WithdrawReasons,
+	},
+	Parameter, StorageDoubleMap, StorageMap, StorageValue,
+};
+use system::ensure_signed;
+
+mod mock;
+mod tests;
+
+pub use self::imbalances::{NegativeImbalance, PositiveImbalance};
+
+pub trait Trait: system::Trait {
+	type Balance: Parameter
+		+ Member
+		+ SimpleArithmetic
+		+ Default
+		+ Copy
+		+ MaybeSerializeDebug + From<u32>;
```

Comment:
```suggestion
		+ MaybeSerializeDebug;
```

`SimpleArithmetic` already implies `From<u32>`